### PR TITLE
Add progress handler to simulation methods

### LINKED ISSUE #315: Add progress handler to `simulate` methods
### 📝 Description of the feature

A progress handler is responsible for displaying progress information to the user. Different scenarios require different progress displays. For instance, an application with a graphical user interface will display progress differently than a command line application.

### 💡 Steps for implementing the feature

Create an IProgress interface class that can be inherited from. IProgress must include a method to update progress. This method should include the following information.
* running state
* percent complete
* context
* error message, if error occurred
* status message, if available

A progress handler parameter needs to be added to `simulate()` and `_simulate()` in methods on the `Additive` object and to `run_simulations()` on the `ParametricStudy` object.

### 🔗 Useful links and references

_No response_

### PR DESCRIPTION
**Pull request by pkrull-ansys**
closes #315

### CODE REVIEW

**Review by safeerehman — APPROVED** [review]

File: src/ansys/additive/core/additive.py [hunk 1]
```diff
@@ -281,6 +287,9 @@ def simulate(
         Microstructure3DInput, list
             Parameters to use for simulations. A list of inputs may be provided to execute multiple
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:288]:
```suggestion
            Parameters to use for simulations. A list of inputs may be provided to run multiple
```

File: src/ansys/additive/core/additive.py [hunk 2]
```diff
@@ -541,24 +560,25 @@ def _simulate_thermal_history(
         input: ThermalHistoryInput,
         out_dir: str,
         server: ServerConnection,
-        logger: ProgressLogger | None = None,
+        progress_handler: IProgressHandler | None = None,
     ) -> ThermalHistorySummary:
         """Execute a thermal history simulation.
```
Comment (julieatansys) [inline src/ansys/additive/core/additive.py:565]:
```suggestion
        """Run a thermal history simulation.
```

File: src/ansys/additive/core/progress_handler.py [hunk 3]
```diff
@@ -19,41 +19,100 @@
 # LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
-"""Provides logging and progress."""
+"""Provides progress handling."""
```
Comment (julieatansys) [inline src/ansys/additive/core/progress_handler.py:22]:
```suggestion
"""Provides progress updates."""
```

File: src/ansys/additive/core/progress_handler.py [hunk 4]
```diff
@@ -19,41 +19,100 @@
...
+            message=progress.message,
+            context=progress.context,
+        )
+
+    def __str__(self):
+        return f"{self.id}: {self.state.name} - {self.percent_complete}% - {self.context} - {self.message}"
+
+
+class IProgressHandler(ABC):
+    """Interface for simulation progress update handlers."""
```
Comment (julieatansys) [inline src/ansys/additive/core/progress_handler.py:72]:
```suggestion
    """Interface for simulation progress updates."""
```

File: src/ansys/additive/core/progress_handler.py [hunk 5]
```diff
@@ -19,41 +19,100 @@
...
+from ansys.api.additive.v0.additive_domain_pb2 import Progress as ProgressMsg
+from ansys.api.additive.v0.additive_domain_pb2 import ProgressState as ProgressMsgState
+from pydantic import BaseModel
 from tqdm import tqdm
 
 
-class ProgressLogger:
-    """Provides for logging and progress reporting."""
+class ProgressState(IntEnum):
+    """Simulation progress state."""
```
Comment (julieatansys) [inline src/ansys/additive/core/progress_handler.py:35]:
```suggestion
    """Simulation progress status."""
```

File: src/ansys/additive/core/progress_handler.py [hunk 6]
```diff
@@ -19,41 +19,100 @@
...
+    """
+
+    def __init__(self):
+        """Initialize progress handler."""
         self._last_percent_complete = 0
         self._last_context = "Initializing"
 
-    def log_progress(self, progress: Progress):
+    def update(self, progress: Progress):
         """Report progress and possibly emit a log message.
```
Comment (julieatansys) [inline src/ansys/additive/core/progress_handler.py:101]: I'm having trouble with "possibly emit." Will a log message be emitted depending on the status? How about: "Report progress and provide a log message as necessary."

Comment (pkrull-ansys) [reply src/ansys/additive/core/progress_handler.py:101]: Good catch! This was a CoPilot generated comment and it's not true. There is no log message emitted.

Comment (pkrull-ansys) [reply src/ansys/additive/core/progress_handler.py]: Updated.

**Review by julieatansys — APPROVED** [review]

**Pull request merged**